docs/docs/client-middleware.md

```diff
@@ -0,0 +1,223 @@
+}
+
+val mwClient = DestinationAttribute(myMiddleware(client), "example")
+```
+```scala mdoc
+mwClient.run(Request[IO](Method.GET, uri"/ok")).use(_.headers.pure[IO]).unsafeRunSync()
+```
+
+## FollowRedirects
+Allows a client to interpret redirect responses and follow them. See [FollowRedirect]
+for configuration.
+
+```scala mdoc:silent
+import org.http4s.client.middleware.FollowRedirect
+val redirectRequest = Request[IO](Method.GET, uri"/redirect")
+```
+```scala mdoc
+client.status(redirectRequest).unsafeRunSync()
+FollowRedirect(maxRedirects = 3)(client).status(redirectRequest).unsafeRunSync()
+```
+
+## GZip
+Adds support for gzip compression. The client will indicate it can read gzip responses
+and if the server responds with gzip, the client will decode the response transparently.
+
+```scala mdoc:silent
+import org.http4s.client.middleware.GZip
+import org.http4s.server.middleware.{GZip => ServerGZip}
+
+val gzipService = ServerGZip(
+  HttpRoutes.of[IO] {
+    case GET -> Root / "long" => Ok("0123456789" * 10) // the body has 100 bytes
+  }
+).orNotFound
+val clientWithoutGzip = Client.fromHttpApp(gzipService)
+val clientWithGzip = GZip()(clientWithoutGzip)
+val longRequest = Request[IO](Method.GET, uri"/long")
+```
+
+```scala mdoc
+// if we request gzip compression the body is smaller
+// than the 100 bytes our route is sending
+clientWithoutGzip
+  .run(longRequest.putHeaders("Accept-Encoding" -> "gzip"))
+  .use(_.body.compile.count).unsafeRunSync()
+// when the client supports gzip we get the uncompressed body
+clientWithGzip.run(longRequest).use(_.body.compile.count).unsafeRunSync()
+```
+
+## Logger, ResponseLogger, RequestLogger
+Log requests and responses. `ResponseLogger` logs the responses, `RequestLogger`
+logs the request, `Logger` logs both.
+
+```scala mdoc:silent
+import org.http4s.client.middleware.Logger
+
+val loggerClient = Logger[IO](
+  logHeaders = false,
+  logBody = true,
+  redactHeadersWhen = _ => false,
+  logAction = Some((msg: String) => Console[IO].println(msg))
+)(client)
+
+```
+```scala mdoc
+loggerClient.expect[Unit](Request[IO](Method.GET, uri"/ok")).unsafeRunSync()
+```
+
+## Retry
+Allows a client to handle server errors by retrying requests. See [Retry] and
+[RetryPolicy] for ways of configuring the retry policy.
+
+```scala mdoc:silent
+import org.http4s.client.middleware.{Retry, RetryPolicy}
+
+object Bail extends Exception("not yet")
+
+// a service that fails the first three times it's called
+val flakyService =
+  Ref[IO].of(0).map { attempts =>
+    HttpRoutes.of[IO] {
+      case _ => attempts.getAndUpdate(_ + 1)
+        .flatMap(a => IO.raiseWhen(a < 3)(Bail) >> Ok("ok"))
+    }
+  }
+
+val policy = RetryPolicy[IO](backoff = _ => Some(1.milli))
+
+```
+```scala mdoc
+flakyService.flatMap { service =>
+    val client = Client.fromHttpApp(service.orNotFound)
+    val retryClient = Retry(policy)(client)
+    retryClient.expect[String](Request[IO](uri = uri"/"))
+} .unsafeRunSync()
```

Comment:
I _really_ like your examples that show the before/after effect of the middleware, I think this could be another good candidate for that.

Reply:
Oh on second thought, that might be confusing with the counter side-effect. Meh :)

Reply:
> Oh on second thought, that might be confusing with the counter side-effect. Meh, this is fine :)

Would another `flakyService.flatMap { service => }` block before, without middleware, be confusing?

Reply:
> > Oh on second thought, that might be confusing with the counter side-effect. Meh, this is fine :)
> 
> Would another `flakyService.flatMap { service => }` block before, without middleware, be confusing?

I did it